File: FastMines_android/src/main/java/fmg/android/app/serializers/MosaicInitDataSerializer.java
package fmg.android.app.serializers;

import android.content.SharedPreferences;

import fmg.common.Logger;
import fmg.common.geom.Matrisize;
import fmg.core.app.serializers.ISerializer;
import fmg.core.app.model.MosaicInitData;
import fmg.core.types.EMosaic;

/** Mosaic data (de)serializer. For save/restore {@link MosaicInitData} */
public class MosaicInitDataSerializer implements ISerializer {

    private static final int VERSION = 1;

    private static final String KEY__MOSAIC_INIT_DATA = MosaicInitData.class.getSimpleName();
    private static final String KEY__MOSAIC_INIT_DATA__VERSION      = KEY__MOSAIC_INIT_DATA + ".version";
    private static final String KEY__MOSAIC_INIT_DATA__SIZE_FIELD_M = KEY__MOSAIC_INIT_DATA + ".sizeField.M";
    private static final String KEY__MOSAIC_INIT_DATA__SIZE_FIELD_N = KEY__MOSAIC_INIT_DATA + ".sizeField.N";
    private static final String KEY__MOSAIC_INIT_DATA__MOSAIC_TYPE  = KEY__MOSAIC_INIT_DATA + ".mosaicType";
    private static final String KEY__MOSAIC_INIT_DATA__COUNT_MINES  = KEY__MOSAIC_INIT_DATA + ".countMines";

    /*
    boolean load(Bundle from, MosaicInitData to) {
        if (from == null)
            return false;

        try {
            int version = from.getInt(KEY__MOSAIC_INIT_DATA__VERSION, 1);
            if (version != 1)
                throw new IllegalArgumentException("Version #" + version + " is not supported");

            MosaicInitData data = new MosaicInitData();
            data.setSizeField(new Matrisize(       from.getInt(KEY__MOSAIC_INIT_DATA__SIZE_FIELD_M, MosaicInitData.DEFAULT_SIZE_FIELD_M),
                                                   from.getInt(KEY__MOSAIC_INIT_DATA__SIZE_FIELD_N, MosaicInitData.DEFAULT_SIZE_FIELD_M)));
            data.setMosaicType(EMosaic.fromOrdinal(from.getInt(KEY__MOSAIC_INIT_DATA__MOSAIC_TYPE , MosaicInitData.DEFAULT_MOSAIC_TYPE.ordinal())));
            data.setCountMines(                    from.getInt(KEY__MOSAIC_INIT_DATA__COUNT_MINES , MosaicInitData.DEFAULT_COUNT_MINES));

            to.copyFrom(data);

            return true;
        } catch(Exception ex) {
            Logger.error("Can not read mosaic init data from Bundle", ex);
            return false;
        }
    }

    void save(MosaicInitData data, Bundle to) {
        to.putInt(KEY__MOSAIC_INIT_DATA__VERSION     , VERSION);
        to.putInt(KEY__MOSAIC_INIT_DATA__SIZE_FIELD_M, data.getSizeField().m);
        to.putInt(KEY__MOSAIC_INIT_DATA__SIZE_FIELD_N, data.getSizeField().n);
        to.putInt(KEY__MOSAIC_INIT_DATA__MOSAIC_TYPE , data.getMosaicType().ordinal());
        to.putInt(KEY__MOSAIC_INIT_DATA__COUNT_MINES , data.getCountMines());
    }
    /**/

    MosaicInitData load(SharedPreferences from) {
        MosaicInitData data = new MosaicInitData();
        if (from != null) try {
            int version = from.getInt(KEY__MOSAIC_INIT_DATA__VERSION, 1);
            if (version != 1)
                throw new IllegalArgumentException("MosaicInitDataSerializer: Version #" + version + " is not supported");

            data.setSizeField(new Matrisize(       from.getInt(KEY__MOSAIC_INIT_DATA__SIZE_FIELD_M, MosaicInitData.DEFAULT_SIZE_FIELD_M),
                                                   from.getInt(KEY__MOSAIC_INIT_DATA__SIZE_FIELD_N, MosaicInitData.DEFAULT_SIZE_FIELD_M)));
            data.setMosaicType(EMosaic.fromOrdinal(from.getInt(KEY__MOSAIC_INIT_DATA__MOSAIC_TYPE , MosaicInitData.DEFAULT_MOSAIC_TYPE.ordinal())));
            data.setCountMines(                    from.getInt(KEY__MOSAIC_INIT_DATA__COUNT_MINES , MosaicInitData.DEFAULT_COUNT_MINES));
        } catch(Exception ex) {
            Logger.error("Can not read mosaic init data from SharedPreferences", ex);
            data = new MosaicInitData(); // reset
        }
        return data;
    }

    void save(MosaicInitData data, SharedPreferences.Editor editor) {
        editor.putInt(KEY__MOSAIC_INIT_DATA__VERSION     , VERSION);
        editor.putInt(KEY__MOSAIC_INIT_DATA__SIZE_FIELD_M, data.getSizeField().m);
        editor.putInt(KEY__MOSAIC_INIT_DATA__SIZE_FIELD_N, data.getSizeField().n);
        editor.putInt(KEY__MOSAIC_INIT_DATA__MOSAIC_TYPE , data.getMosaicType().ordinal());
        editor.putInt(KEY__MOSAIC_INIT_DATA__COUNT_MINES , data.getCountMines());
    }

}

File: FastMines_uwp/src/fmg/uwp/app/Serializers/MosaicInitDataSerializer.cs
using System;
using Windows.Foundation.Collections;
using Fmg.Common;
using Fmg.Common.Geom;
using Fmg.Core.App.Model;
using Fmg.Core.App.Serializers;
using Fmg.Core.Types;

namespace Fmg.Uwp.App.Serializers {

    /// <summary> Mosaic data (de)serializer. For save/restore <see cref="MosaicInitData"/> </summary>
    public class MosaicInitDataSerializer : ISerializer {

        private const int VERSION = 1;

        private const           string KEY__MOSAIC_INIT_DATA__SECTION      = nameof(MosaicInitData);
        private const           string KEY__MOSAIC_INIT_DATA__VERSION      = ".version";
        private static readonly string KEY__MOSAIC_INIT_DATA__SIZE_FIELD_M = nameof(MosaicInitData.SizeField) + '.' + nameof(Matrisize.m);
        private static readonly string KEY__MOSAIC_INIT_DATA__SIZE_FIELD_N = nameof(MosaicInitData.SizeField) + '.' + nameof(Matrisize.n);
        private const           string KEY__MOSAIC_INIT_DATA__MOSAIC_TYPE  = nameof(MosaicInitData.MosaicType);
        private const           string KEY__MOSAIC_INIT_DATA__COUNT_MINES  = nameof(MosaicInitData.CountMines);

        public MosaicInitData Load(IPropertySet from) {
            if (from.ContainsKey(KEY__MOSAIC_INIT_DATA__SECTION)) try {
                var section = (Windows.Storage.ApplicationDataCompositeValue)from[KEY__MOSAIC_INIT_DATA__SECTION];
            
                if (!section.ContainsKey(KEY__MOSAIC_INIT_DATA__VERSION))
                    return new MosaicInitData();

                int version = (int)section[KEY__MOSAIC_INIT_DATA__VERSION];
                if (version != 1)
                    throw new ArgumentException("MosaicInitDataSerializer: Version #" + version + " is not supported");

                int mosaicTypeOrdinal = (int)section[KEY__MOSAIC_INIT_DATA__MOSAIC_TYPE];
                int sizeFieldM        = (int)section[KEY__MOSAIC_INIT_DATA__SIZE_FIELD_M];
                int sizeFieldN        = (int)section[KEY__MOSAIC_INIT_DATA__SIZE_FIELD_N];
                int countMines        = (int)section[KEY__MOSAIC_INIT_DATA__COUNT_MINES];

                return new MosaicInitData {
                    MosaicType = EMosaicEx.FromOrdinal(mosaicTypeOrdinal),
                    SizeField  = new Matrisize(sizeFieldM, sizeFieldN),
                    CountMines = countMines
                };
            } catch(Exception ex) {
                Logger.Error($"Can`t load mosaic init data: {ex.Message}");
            }
            return new MosaicInitData();
        }

        public void Save(MosaicInitData data, IPropertySet to) {
            to[KEY__MOSAIC_INIT_DATA__SECTION] = new Windows.Storage.ApplicationDataCompositeValue() {
                [KEY__MOSAIC_INIT_DATA__VERSION     ] = VERSION,
                [KEY__MOSAIC_INIT_DATA__MOSAIC_TYPE ] = data.MosaicType.Ordinal(),
                [KEY__MOSAIC_INIT_DATA__SIZE_FIELD_M] = data.SizeField.m,
                [KEY__MOSAIC_INIT_DATA__SIZE_FIELD_N] = data.SizeField.n,
                [KEY__MOSAIC_INIT_DATA__COUNT_MINES ] = data.CountMines,
            };
        }

    }

}
